cachelib/cachebench/workload/BinaryKVReplayGenerator.h

```diff
@@ -0,0 +1,289 @@
+    XCHECK_LT(shardId, numShards_);
+    return *stressorCtxs_[shardId];
+  }
+
+  inline StressorCtx& getStressorCtx() {
+    if (!stressorIdx_.get()) {
+      stressorIdx_.reset(new uint32_t(incrementalIdx_++));
+    }
+
+    return getStressorCtx(*stressorIdx_);
+  }
+
+  inline Request* getRequestPtr(size_t shardId) {
+    XCHECK_LT(shardId, numShards_);
+    return requestPtr_[shardId];
+  }
+
+  inline Request* getRequestPtr() {
+    if (!stressorIdx_.get()) {
+      stressorIdx_.reset(new uint32_t(incrementalIdx_++));
+    }
+
+    return getRequestPtr(*stressorIdx_);
+  }
+};
+
+inline void BinaryKVReplayGenerator::genRequests(folly::Latch& latch) {
+  bool init = true;
+  uint64_t nreqs = 0;
+  binaryStream_.setOffset(fastForwardCount_);
+  auto begin = util::getCurrentTimeSec();
+  while (!shouldShutdown()) {
+    try {
+      BinaryRequest* req = binaryStream_.getNextPtr();
+      auto key = req->getKey(binaryStream_.getKeyOffset());
+      XDCHECK_LT(req->op_, 11);
+      nreqs++;
+      auto shardId = getShard(key);
+      auto& stressorCtx = getStressorCtx(shardId);
+      auto& reqQ = *stressorCtx.reqQueue_;
+      while (!reqQ.write(req) && !stressorCtx.isFinished() &&
+             !shouldShutdown()) {
+        // ProducerConsumerQueue does not support blocking, so use sleep
+        if (init) {
+          latch.count_down();
+          init = false;
+        }
+        std::this_thread::sleep_for(
+            std::chrono::microseconds{checkIntervalUs_});
+      }
+      if (nreqs >= preLoadReqs_ && init) {
+        auto end = util::getCurrentTimeSec();
+        double reqsPerSec = nreqs / (double)(end - begin);
+        XLOGF(INFO, "Parse rate: {:.2f} reqs/sec", reqsPerSec);
+        latch.count_down();
+        init = false;
+      }
+    } catch (const EndOfTrace& e) {
+      if (init) {
+        latch.count_down();
+      }
+      break;
+    }
+  }
+
+  setEOF();
+}
+
+const Request& BinaryKVReplayGenerator::getReq(uint8_t,
+                                               std::mt19937_64&,
+                                               std::optional<uint64_t>) {
+  BinaryRequest* req = nullptr;
+  auto& stressorCtx = getStressorCtx();
+  auto& reqQ = *stressorCtx.reqQueue_;
+  auto& resubmitQueue = stressorCtx.resubmitQueue_;
+  XDCHECK_EQ(req, nullptr);
+
+  while (resubmitQueue.empty() && !reqQ.read(req)) {
+    if (resubmitQueue.empty() && isEOF()) {
+      throw cachelib::cachebench::EndOfTrace("Test stopped or EOF reached");
+    }
+    // ProducerConsumerQueue does not support blocking, so use sleep
+    std::this_thread::sleep_for(std::chrono::microseconds{checkIntervalUs_});
+  }
+
+  if (req == nullptr) {
+    XCHECK(!resubmitQueue.empty());
+    req = resubmitQueue.front();
+    resubmitQueue.pop();
+  }
+  XDCHECK_NE(req, nullptr);
+  XDCHECK_NE(reinterpret_cast<uint64_t>(req), 0);
+  XDCHECK_LT(req->op_, 12);
+  auto key = req->getKey(binaryStream_.getKeyOffset());
+  OpType op;
+  switch (req->op_) {
```

Comment:
you can just cast OptType to uint8_t here and avoid the switch

Reply:
done